convert2rhel/unit_tests/redhatrelease_test.py

```diff
@@ -142,13 +158,30 @@ def test_yum_patch(monkeypatch, modified, caplog):
         assert "Skipping patching, yum configuration file not modified" in caplog.text
 
 
-YUM_CONF_WITHOUT_DISTROVERPKG = """[main]
-installonly_limit=3
+@unit_tests.mock(os.path, "isfile", lambda x: True)
+def test_get_os_release_filepath(caplog):
+    assert get_os_release_filepath() == "/etc/os-release"
 
-#  This is the default"""
+    os.path.isfile = lambda x: False
```

Comment:
I would make a separate test for this case.  My rule of thumb (but it's only from my experience, let me know if you have a better guideline) is that you can assert multiple times in a single unittest but you should only call the function being tested a single time.  (There are exceptions to this, like if you're testing idempotence where you want to make sure that the state of the system does not change after the function is called a second time.  But I don't think that applies here.)

Reply:
@abadger generally, I follow the same rule (always separate the good from the bad tests), but for this one, I just followed the convention of other tests in the codebase. The only place I saw using the separate good/bad tests was on class tests, which I believe is not the case on this one. I kinda like the way of using good/bad tests in the `pytest.mark.parametrize`, it's concise and clean, and if you have to add new tests cases for that, it's not that difficult after all.

Reply:
We try to call the functions under test just once in a test. In case you need to catch the Exception, we typically do it the way you have it now - calling it either with or without the `with pytest.raises(<Exc>):` statement, depending on a condition provided by `parametrize`. So it's not called twice within one test run.

We started using `parametrize` quite late so you can see that some functions under test are tested by multiple test functions instead of just one test function with `parametrize`. I myself like the parametrize approach more as it prevents code duplication.

@abadger, let us know if you know of a more elegant solution.

Reply:
I love parametrize too!  However, I love it so much that I have had colleagues point out that it can easily lead to tests that are hard to read and debug (it makes what the test does declarative but that also means that reasoning about what a test is doing [to debug it or to add a new test case] is hard to follow and the relevant information is located in multiple locations).  I haven't completely cured myself of overusing parametrization but there are a few rules that seem to work well:

* Use a new test if it would require adding a conditional to the current test's code.
* Use a new test if it you'd have to create different mock objects for it in the current test.
* Use a new test if the output that you check for has to follow a different pattern (type, format string, regex, exception, etc) for the new parameter.

Over parametrization often goes along with this antipattern (story format): https://www.informit.com/articles/article.aspx?p=1392524  (more traditional): https://martinfowler.com/bliki/FlagArgument.html  Instead of creating a different function whose name can clue you in on what is happening, you end up with a long data structure of True|False values that aren't obvious on their own what is happening.  Looking up the meaning of the values is possible but has a cost when debugging or adding a new test case.

Mentioning debugging brings up a difference (in emphasis) that I see between test code and regular code.  All code should be as simple to read as possible, but it's especially important for test code.  When your unit tests fail, you want it to be obvious what the test case is doing.  It's not a good feeling to have a failing test and spend most of the time trying to decide if the test case is wrong.

For the question of preventing code duplication, I think that fixtures are what's appropriate in the general case (and parametrization is appropriate in specific cases).  pytest fixtures are very flexible and you can push large sections of boilerplate into them (although, again, be careful that you don't make it so someone has to spend more time determining if your test is correct than looking for the bug in their code).

I think it's debatable whether duplication is appropriate to worry about here as there's only two lines in common but to see what it looks like:

``` python
@pytest.fixture    
def monkeypatch_isfile(request, monkeypatch):    
    isfile_returns = request.node.get_closest_marker("isfile_returns")    
    if isfile_returns is None:    
        isfile_returns = False    
    else:    
        isfile_returns = isfile_returns.args[0]    
    
    is_file_mock = mock.MagicMock(return_value=isfile_returns)    
    monkeypatch.setattr(os.path, "isfile", value=is_file_mock)    
    return is_file_mock    

@pytest.mark.isfile_returns(True)    
def test_get_os_release_filepath(monkeypatch_isfile, caplog):    
    assert get_os_release_filepath() == "/etc/os-release"    
    
@pytest.mark.isfile_returns(False)    
def test_get_os_release_filepath_file_not_found(monkeypatch_isfile, caplog):    
    with pytest.raises(SystemExit):    
        get_os_release_filepath()    
    assert (    
        "Error: Unable to find the /etc/os-release file containing the information needed."    
        in caplog.records[-1].message    
    )
```

Reply:
Great write up, @abadger. I agree with it in full. Some duplication is indeed better than a test that hard to decipher. The fixture approach is something I haven't considered and I like it.

The Clean Code Tip story is great. I fear our codebase that smells like it a bit on couple of places.

Reply:
I 'd say the tests as @r0x0d has them now are short and easily readable/understandable. Or would you like him to change them somehow?

Reply:
It is okay with me.  I would like to see good and bad test cases split up in the future but I think maybe that could be a task I take on for all the tests in the future?  Would you be okay with that sort of mass cleanup from me?

Reply:
Sure, I'm definitely not against that.